### PR TITLE
feat: Support Reactive Routes

### LINKED ISSUE #1114: Support Reactive Routes
It would be nice if we had the same behavior as JAX-RS methods when using https://quarkus.io/guides/reactive-routes

### PR DESCRIPTION
**Pull request by angelozerr**
feat: Support Reactive Routes

Fixes #1114

### CODE REVIEW

**Comment by fbricon** [comment]
we should have unit tests for that

**Comment by angelozerr** [comment]
> we should have unit tests for that

Yes sure. I have implemented quickly the support and now I need to clean my code and write tests.

@gastaldi here the result:

 * classic usecase:

image

 * routebase usecase:
 
image

 * mutiple route usecase:
 
image

**Comment by angelozerr** [comment]
My code is clean now, I need to write tests but @fbricon @gastaldi you can start to review it

**Comment by fbricon** [comment]
> * mutiple route usecase:
> 
> image

I wonder if we should have 1 URL per line. How many different routes can a user add, realistically? @gastaldi WDYT?

**Comment by gastaldi** [comment]
@fbricon in my experience, a single `@Route` is enough per method

File: src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/utils/AnnotationUtils.java [hunk 1]
```diff
@@ -97,6 +99,28 @@ private static PsiAnnotation getFirstAnnotation(PsiAnnotation[] annotations, Str
...
+		List<PsiAnnotation> all = new ArrayList<>();
+		if (annotatable instanceof PsiAnnotationOwner) {
+			collectAnnotation(((PsiAnnotationOwner) annotatable).getAnnotations(), all, annotationNames);
+		} else if (annotatable instanceof PsiModifierListOwner) {
+			collectAnnotation(((PsiModifierListOwner) annotatable).getAnnotations(), all, annotationNames);
+		}
+		return all;
+	}
+
+	private static void collectAnnotation(PsiAnnotation[] annotations, List<PsiAnnotation> all, String...annotationNames) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/utils/AnnotationUtils.java:112]: collectAnnotations

Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/utils/AnnotationUtils.java]: fixed

**Comment by angelozerr** [comment]
I support now impicit usage when a methode is annotated with Route:

image

**Comment by angelozerr** [comment]
> I wonder if we should have 1 URL per line. How many different routes can a user add, realistically? @gastaldi WDYT?

I fear that we takes some times to do implement that for few usages. if you want to have that really, we could do that in a separate PR.

**Comment by angelozerr** [comment]
Now tests exist.

File: src/main/java/com/redhat/microprofile/psi/internal/quarkus/route/java/ReactiveRouteImplicitUsageProvider.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    public boolean isImplicitUsage(@NotNull PsiElement element) {
+        return isImplicitRead(element) || isImplicitWrite(element);
+    }
+
+    @Override
+    public boolean isImplicitRead(@NotNull PsiElement element) {
+        if (element instanceof PsiMethod) {
+            PsiMethod method = (PsiMethod)element;
+            if (isReactiveRoute(method)) {
```
Comment (fbricon) [inline src/main/java/com/redhat/microprofile/psi/internal/quarkus/route/java/ReactiveRouteImplicitUsageProvider.java:43]: return isReactiveRoute(method)

Comment (angelozerr) [reply src/main/java/com/redhat/microprofile/psi/internal/quarkus/route/java/ReactiveRouteImplicitUsageProvider.java]: fixed

File: src/main/java/com/redhat/microprofile/psi/internal/quarkus/route/java/ReactiveRouteConstants.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public static final String REACTIVE_ROUTE_GET_ANNOTATION = "io.quarkus.vertx.web.Route.HttpMethod.GET";
+
+    public static final String REACTIVE_ROUTE_HEAD_ANNOTATION = "io.quarkus.vertx.web.Route.HttpMethod.HEAD";
+
+    public static final String REACTIVE_ROUTE_POST_ANNOTATION = "io.quarkus.vertx.web.Route.HttpMethod.POST";
+
+    public static final String REACTIVE_ROUTE_PUT_ANNOTATION = "io.quarkus.vertx.web.Route.HttpMethod.PUT";
+
+    public static final String REACTIVE_ROUTE_DELETE_ANNOTATION = "io.quarkus.vertx.web.Route.HttpMethod.DELETE";
```
Comment (fbricon) [inline src/main/java/com/redhat/microprofile/psi/internal/quarkus/route/java/ReactiveRouteConstants.java:43]: Seem unused constants

Comment (angelozerr) [reply src/main/java/com/redhat/microprofile/psi/internal/quarkus/route/java/ReactiveRouteConstants.java]: Good catch! removed.

**Comment by fbricon** [comment]
Thanks @angelozerr

**Pull request merged**